File: 2019-20/contest4/prog/intermediate/patolli.java
import java.io.*;
import java.lang.*;
import java.util.*;

public class patolli {
	static public boolean isPrime(int num) {
		if(num < 2)
			return false;
		int end = (int)Math.sqrt(num);
		for(int n = 2; n <= end; ++n)
			if( (num / n) * n == num ) // divisible check
				return false;
		return true;
	}

	public static void setMarker(int curr, int next, int[] players) {
		for(int i = 0; i < players.length; ++i) {
			if(curr == players[i]) {
				players[i] = next;
				break;
			}
		}
	}

	public static boolean occupied(int curr, int next, int[] players) {
		for(int i = 0; i < players.length; ++i) {
			if(curr == players[i])
				continue;
			if(next == players[i])
				return true;
		}
		return false;
	}

	public static void main(String[] args) {
		Scanner input = new Scanner(System.in);
		for(int n = 0; n < 5; ++n) {
			int[] opponent = new int[] { input.nextInt(), input.nextInt(), input.nextInt() };
			int[] players = new int[] { input.nextInt(), input.nextInt(), input.nextInt() };
			int num_rolls = input.nextInt();
			int[] rolls = new int[num_rolls];
			for(int i = 0; i < num_rolls; ++i)
				rolls[i] = input.nextInt();
			for(int r = 0; r < num_rolls; ++r) {
				int curr = Math.min(Math.min(players[0],players[1]),players[2]);
				if(curr == 52)
					break;
				int roll = rolls[r];
				int next = curr + roll;
				// Check if will land on an occupied location
				if( next == opponent[0] || next == opponent[1] || next == opponent[2] || occupied(curr, next, players) )
					continue;
				// Check if will land on target location 52
				if( next == 52 ) {
					setMarker(curr, next, players);
					break;
				}
				// Check if will land beyond 52
				if( next > 52 )
					continue;
				// Check if will land on prime #
				if( isPrime(next) ) {
					for(int i = 0; i < 6; ++i) {
						++next;
						if( next == opponent[0] || next == opponent[1] || next == opponent[2] || occupied(curr, next, players) ) {
							--next;
							break;
						}
					}
					setMarker(curr, next, players);
					continue;
				}
				// Check if will land on perfect square > 4
				if( next == 9 || next == 16 || next == 25 || next == 36 || next == 49 ) {
					for(int i = 0; i < 6; ++i) {
						--next;
						if( next == opponent[0] || next == opponent[1] || next == opponent[2] || occupied(curr, next, players) ) {
							++next;
							break;
						}
					}
					setMarker(curr, next, players);
					continue;
				}
				// Check if one horizontal move followed by one vertical move
				if( (curr <  7 &&  7 < next) ||
					(curr < 12 && 12 < next) ||
					(curr < 17 && 17 < next) ||
					(curr < 22 && 22 < next) ||
					(curr < 27 && 27 < next) ||
					(curr < 35 && 35 < next) ||
					(curr < 40 && 40 < next) ||
					(curr < 45 && 45 < next) ||
					(curr < 50 && 50 < next) )
				{
					for(; next > curr; --next) {
						if( (next / roll) * roll == next && next != opponent[0] && next != opponent[1] && next != opponent[2] && !occupied(curr, next, players) )
							// Unoccupied multiple of roll
							break;
					}
					setMarker(curr, next, players);
					continue;
				}
				setMarker(curr, next, players);
			}
			Arrays.sort(players);
			if(players[0] == 52 && players[1] == 52 && players[2] == 52)
				System.out.println("GAME OVER");
			else {
				for(int player : players)
					if(player == 52)
						break;
					else {
						System.out.print(player);
						System.out.print(' ');
					}
				System.out.println();
			}
		}
	}
}

File: 2019-20/contest4/prog/intermediate/patolli.py
import sys
import math

def isPrime(num):
	if num < 2:
		return False
	for n in range(2,int(math.sqrt(num))+1):
		if (num // n) * n == num: # divisible check
			return False
	return True

def setMarker(curr, next, players):
	for i in range(len(players)):
		if players[i] == curr:
			players[i] = next
			return
	return

def occupied(curr, next, players):
	for player in players:
		if curr == player:
			continue
		if next == player:
			return True
	return False

if __name__== "__main__":
	for n in range(5):
		nums = [int(x) for x in input().split()]
		opponent = nums[:3]
		players = nums[3:6]
		num_rolls = nums[6]
		rolls = nums[7:]
		for k in range(num_rolls):
			curr = min(players)
			if curr == 52:
				break
			roll = rolls[k]
			next = curr + roll
			#print("curr="+str(curr)+",roll="+str(roll)+",next="+str(next))
			# Check if will land on an occupied location
			if next == opponent[0] or next == opponent[1] or next == opponent[2] or occupied(curr, next, players):
				continue
			# Check if will land on target location 52
			if next == 52:
				setMarker(curr, next, players)
				break
			# Check if will land beyond 52
			if next > 52:
				continue
			# Check if will land on a prime number
			if isPrime(next):
				for i in range(6):
					next += 1
					if next == opponent[0] or next == opponent[1] or next == opponent[2] or occupied(curr, next, players):
						next -= 1
						break
				setMarker(curr, next, players)
				continue
			# Check if will land on perfect square > 4
			if next == 9 or next == 16 or next == 25 or next == 36 or next == 49:
				for i in range(6):
					next -= 1
					if next == opponent[0] or next == opponent[1] or next == opponent[2] or occupied(curr, next, players):
						next += 1
						break
				setMarker(curr, next, players)
				continue
			# Check if one horizontal move followed by one vertical move
			if (curr <  7 and  7 < next) or (curr < 12 and 12 < next) or (curr < 17 and 17 < next) or (curr < 22 and 22 < next) or (curr < 27 and 27 < next) or (curr < 35 and 35 < next) or (curr < 40 and 40 < next) or (curr < 45 and 45 < next) or (curr < 50 and 50 < next):
				while next > curr:
					if (next // roll) * roll == next and next != opponent[0] and next != opponent[1] and next != opponent[2] and not occupied(curr, next, players):
						break
					next -= 1
				setMarker(curr, next, players)
				continue
			setMarker(curr, next, players)
		players.sort()
		if players[0] == 52 and players[1] == 52 and players[2] == 52:
			print("GAME OVER")
		else:
			for player in players:
				if player == 52:
					break
				print(player, end = ' ')
			print()
